### PR TITLE
Toronto: Add councillor org membership from councillor page

### PR DESCRIPTION
**Pull request by patcon**
https://github.com/patcon/scrapers-ca/commit/dd0c7d7a51d6dea338b19ff3ae2035ff08cb207a

Pending merge of https://github.com/opencivicdata/pupa/pull/236

### CODE REVIEW

**Comment by patcon** [comment]
Ok, this is waaaaaay better. We can fetch committee membership by meeting, so I just hardcoded a dict of two-letter committee codes and reference meeting ids, and we use that to scrape all members. Right now, we're using the fact that councillors are linked to only add membership info for councillors.

Later, we could decide to add non-councillor persons to get more data into OCD.

We could also add membership data for all the other committees, as we're just scraping standing committees right now.

This is fine for merge, and I'll submit those changes in other PRs.

**Comment by patcon** [comment]
Hm. We've had a name change on council, so this won't work until I submit another tiny PR

**Comment by patcon** [comment]
Rebased this over #183

File: ca_on_toronto/committees.py [hunk 1]
```diff
@@ -4,8 +4,56 @@
...
+    def allMembers(self, member_list_url):
+        members = []
+
+        page = self.lxmlize(member_list_url)
+        for li in page.xpath('//ul/li'):
+            member = {'is_councillor': False}
+
+            if li.xpath('.//a'): member['is_councillor'] = True
+
+            li_re = re.compile(r'^(?P<name>.+?)(?: \((?P<role>.+)\))?$')
```
Comment (jpmckinney) [inline ca_on_toronto/committees.py]: The regex should be compiled outside the loop, no?

File: ca_on_toronto/committees.py [hunk 2]
```diff
@@ -4,8 +4,56 @@
 from .helpers import build_lookup_dict, committees_from_sessions
 from .constants import TWO_LETTER_ORG_CODE_SCHEME
 
+import re
+
+
+MEMBERS_URL_TEMPLATE = 'http://app.toronto.ca/tmmis/decisionBodyProfile.do?function=doGetMembers&meetingId={}&showLink=true'
+DEFAULT_COMMITTEE_ROLE = 'Member'
+
+REFERENCE_MEETING_IDS = {
```
Comment (jpmckinney) [inline ca_on_toronto/committees.py]: Doesn't this basically freeze committee members (until you update the meeting IDs)? Or is a meeting ID really a committee ID? If it's a meeting, then the corresponding web page is static.

File: ca_on_toronto/committees.py [hunk 3]
```diff
@@ -4,8 +4,56 @@
 from .helpers import build_lookup_dict, committees_from_sessions
 from .constants import TWO_LETTER_ORG_CODE_SCHEME
 
+import re
+
+
+MEMBERS_URL_TEMPLATE = 'http://app.toronto.ca/tmmis/decisionBodyProfile.do?function=doGetMembers&meetingId={}&showLink=true'
+DEFAULT_COMMITTEE_ROLE = 'Member'
```
Comment (jpmckinney) [inline ca_on_toronto/committees.py]: I think OCD uses lowercase.

File: ca_on_toronto/committees.py [hunk 4]
```diff
@@ -4,8 +4,56 @@
...
 
 class TorontoCommitteeScraper(CanadianScraper):
+    def allMembers(self, member_list_url):
+        members = []
+
+        page = self.lxmlize(member_list_url)
+        for li in page.xpath('//ul/li'):
+            member = {'is_councillor': False}
+
+            if li.xpath('.//a'): member['is_councillor'] = True
```
Comment (jpmckinney) [inline ca_on_toronto/committees.py]: Can you document in the code why members wouldn't be councillors?

File: ca_on_toronto/committees.py [hunk 5]
```diff
@@ -4,8 +4,56 @@
...
+            member['name'] = matches.group('name').strip()
+
+            yield member
+
+    def councillorMembers(self, org_code):
+        ref_meeting_id = REFERENCE_MEETING_IDS.get(org_code)
+        if ref_meeting_id:
+            membership_url = MEMBERS_URL_TEMPLATE.format(ref_meeting_id)
+            for member in self.allMembers(membership_url):
+                if member['is_councillor']: yield member
```
Comment (jpmckinney) [inline ca_on_toronto/committees.py]: Never do one-line if-statements. Makes the code harder to skim.

File: ca_on_toronto/committees.py [hunk 6]
```diff
@@ -25,6 +73,12 @@ def scrape(self):
                     o = Organization(name=inst['name'], classification='committee')
                     extras.update({'description': inst['info']})
                     o.add_identifier(inst['code'], scheme=TWO_LETTER_ORG_CODE_SCHEME)
+
+                    print(inst['code'])
```
Comment (jpmckinney) [inline ca_on_toronto/committees.py]: Remove print.

File: ca_on_toronto/committees.py (hunk 2 above)
Comment (patcon) [inline ca_on_toronto/committees.py]: Yeah, it's not ideal, but it works for this term, and I can foresee how to improve it once the end of the term nears.

Basically:
* members by decisionBodyId are *sometimes* available, but not always, but lack the role information: (I also can't figure out the rhyme or reason for when this links works and when it 404s)
http://app.toronto.ca/tmmis/decisionBodyProfile.do?function=doGetMembers&decisionBodyId=1032&showLink=true
* meetingId seems to have more complete information:
http://app.toronto.ca/tmmis/decisionBodyProfile.do?function=doGetMembers&meetingId=11014&showLink=true
* but meetingId urls vary in time -- you can see this by expanding meeting links here:http://app.toronto.ca/tmmis/decisionBodyProfile.do?function=doPrepare&decisionBodyId=1181#Meeting-2016.AU5 (future dates seem to lack role data, and past dates sometimes lack persons or roles, almost like it's a reflection of attendance)

So basically, it seems attendance data is inconsistent in ways that no one has clarified to me yet. And so I was instead opting for a representative meeting.

In the future, we could add a means to request each meeting page (app.toronto.ca/tmmis/decisionBodyProfile.do?function=doPrepare&decisionBodyId=1181) and follow the redirect in order to parse the most recent meeting from the fragment identifier (http://app.toronto.ca/tmmis/decisionBodyProfile.do?function=doPrepare&decisionBodyId=1181#Meeting-2016.AU5), and then hit that link. But it seemed a bit more complicated, and so this seemed a better route for now?

File: ca_on_toronto/committees.py (hunk 3 above)
Comment (patcon) [inline ca_on_toronto/committees.py]: Actually, not seeing that in NYC or Chicago data:

* http://ocd.datamade.us/ocd-person/b998bc09-5e16-4117-a93a-e25d41b826a2/
* https://github.com/datamade/chi-councilmatic/blob/master/councilmatic/settings_jurisdiction.py#L98-L102

I'd prefer to stay consistent so I don't have to get title-case changes into django-councilmatic

File: ca_on_toronto/committees.py (hunk 4 above)
Comment (patcon) [inline ca_on_toronto/committees.py]: Done and done

File: ca_on_toronto/committees.py (hunk 5 above)
Comment (patcon) [inline ca_on_toronto/committees.py]: Makes sense. Done

File: ca_on_toronto/committees.py (hunk 6 above)
Comment (patcon) [inline ca_on_toronto/committees.py]: Done

File: ca_on_toronto/committees.py (hunk 1 above)
Comment (patcon) [inline ca_on_toronto/committees.py]: Ah my bad. thanks

File: ca_on_toronto/committees.py (hunk 2 above)
Comment (patcon) [inline ca_on_toronto/committees.py]: Actually, I can add a "term" layer to the hash. that's an easy way to address one of your points. one sec

**Comment by patcon** [comment]
Made some changes (no history rewriting) and this is good for another review.

cc: @shaqsingh (there were some recent comments on your #173, but I had to fix the scraper in @4be349dae1f4f90b74490de607239afc58ea9d04 to get this working :/)

File: ca_on_toronto/committees.py [hunk 7]
```diff
@@ -1,11 +1,83 @@
...
 from .helpers import build_lookup_dict, committees_from_sessions
 from .constants import TWO_LETTER_ORG_CODE_SCHEME
 
+import re
+
+
+MEMBERSHIP_URL_TEMPLATE = 'http://app.toronto.ca/tmmis/decisionBodyProfile.do?function=doGetMembers&meetingId={}&showLink=true'
+DEFAULT_COMMITTEE_ROLE = 'Member'
+
+REFERENCE_MEETING_IDS = defaultdict(lambda: {})
```
Comment (jpmckinney) [inline ca_on_toronto/committees.py]: `defaultdict(dict)` is more canonical.

File: ca_on_toronto/committees.py [hunk 8]
```diff
@@ -1,11 +1,83 @@
...
+        * name (string)
+        * role (string)
+        * is_councillor (bool)
+        """
+        members = []
+
+        page = self.lxmlize(member_list_url)
+        li_re = re.compile(r'^(?P<name>.+?)(?: \((?P<role>.+)\))?$')
+        for li in page.xpath('//ul/li'):
+            member = {'is_councillor': False}
```
Comment (jpmckinney) [inline ca_on_toronto/committees.py]: Btw, you can just do `bool(li.xpath('.//a'))` here.

File: ca_on_toronto/committees.py [hunk 9]
```diff
@@ -1,11 +1,83 @@
...
 from .constants import TWO_LETTER_ORG_CODE_SCHEME
 
+import re
+
+
+MEMBERSHIP_URL_TEMPLATE = 'http://app.toronto.ca/tmmis/decisionBodyProfile.do?function=doGetMembers&meetingId={}&showLink=true'
+DEFAULT_COMMITTEE_ROLE = 'Member'
+
+REFERENCE_MEETING_IDS = defaultdict(lambda: {})
+REFERENCE_MEETING_IDS['2014-2018'] = {
```
Comment (jpmckinney) [inline ca_on_toronto/committees.py]: Can you document how these IDs are chosen?

File: ca_on_toronto/committees.py (hunk 7 above)
Comment (patcon) [inline ca_on_toronto/committees.py]: :+1:

**Comment by jpmckinney** [comment]
A few more comments, but looks good.

File: ca_on_toronto/committees.py (hunk 8 above)
Comment (patcon) [inline ca_on_toronto/committees.py]: :+1:

File: ca_on_toronto/committees.py (hunk 9 above)
Comment (patcon) [inline ca_on_toronto/committees.py]: :+1:

**Comment by patcon** [comment]
Finito.

As for the meetingIds, I'm eager to sort out a better way to get that membership data, but the city site means this will be quite hack-ish, and I'd love to have something simple and hardcoded working for now :)
